libs/payment/api/src/lib/payment-provider/payrexx-payment-provider.ts

```diff
@@ -5,170 +5,237 @@ import {
+    const amount = invoice.items.reduce(
+      (accumulator, {amount, quantity}) => accumulator + amount * quantity,
+      0
+    )
+    const transaction = await this.transactionClient.chargePreAuthorizedTransaction(
+      parseInt(customerID),
+      {
+        amount,
+        referenceId: paymentID
+      }
+    )
+    return {
+      intentID: transaction.id.toString(),
+      intentSecret: successURL,
+      intentData: JSON.stringify(transaction),
+      state: PaymentState.submitted
+    }
   }
 
-  async createIntent(props: CreatePaymentIntentProps): Promise<Intent> {
-    const data = {
+  private async createGatewayIntent({
+    invoice,
+    paymentID,
+    successURL,
+    failureURL
+  }: CreatePaymentIntentProps) {
+    const amount = invoice.items.reduce(
+      (accumulator, {amount, quantity}) => accumulator + amount * quantity,
+      0
+    )
+    const gateway = await this.gatewayClient.createGateway({
       psp: this.psp,
       pm: this.pm,
-      referenceId: props.paymentID,
-      amount: props.invoice.items.reduce(
-        (prevItem, currentItem) => prevItem + currentItem.amount * currentItem.quantity,
-        0
-      ),
+      referenceId: paymentID,
+      amount,
       fields: {
         email: {
-          value: props.invoice.mail
+          value: invoice.mail
         }
       },
-      successRedirectUrl: props.successURL,
-      failedRedirectUrl: props.failureURL,
-      cancelRedirectUrl: props.failureURL,
+      successRedirectUrl: successURL,
+      failedRedirectUrl: failureURL,
+      cancelRedirectUrl: failureURL,
       vatRate: this.vatRate,
-      currency: 'CHF'
-    }
-    const signature = crypto
-      .createHmac('sha256', this.instanceAPISecret)
-      .update(qs.stringify(data))
-      .digest('base64')
-
-    const res = await fetch(
-      `https://api.payrexx.com/v1.0/Gateway/?instance=${encodeURIComponent(this.instanceName)}`,
-      {
-        method: 'POST',
-        body: qs.stringify({...data, ApiSignature: signature})
-      }
-    )
-    if (res.status !== 200) throw new Error(`Payrexx response is NOK with status ${res.status}`)
-    const payrexxResponse = (await res.json()) as PayrexxResponse
-
+      currency: 'CHF',
+      preAuthorization: true,
+      chargeOnAuthorization: true
+    })
     logger('payrexxPaymentProvider').info(
       'Created Payrexx intent with ID: %s for paymentProvider %s',
-      payrexxResponse.data?.[0].id,
+      gateway.id,
       this.id
     )
-    // in case Payrexx throws an error
-    if (payrexxResponse.status === 'error') {
-      throw new Error(`Error from Payrexx: ${payrexxResponse.message}`)
-    }
-
     return {
-      intentID: payrexxResponse.data?.[0].id,
-      intentSecret: payrexxResponse.data?.[0].link,
-      intentData: JSON.stringify(payrexxResponse.data),
+      intentID: gateway.id.toString(),
+      intentSecret: gateway.link,
+      intentData: JSON.stringify(gateway),
       state: PaymentState.submitted
```

Comment:
no need to return preauthorizationId aka customerID here?

Reply:
There is no customerId here in create gateway response.
This is creation of the gateway. Only after the gateway is paid we can check gateway transaction or use webhook information. Which we do.